### PR TITLE
Add Pretendo Support

### LINKED ISSUE #10: Initial support for Third-Party Nintendo Network servers : Pretendo
As you probably know, the Nintendo Network servers for the 3DS, won’t be around for a long time, but the community is at work to still keep it alive, by archiving content, and hosting their Nintendo Network server.

[https://pretendo.network](https://pretendo.network)

Pretendo is just that, and is currently aiming at restoring support for all Nintendo Network features for both platforms, app & game wise (Including, already removed things like the Miiverse). While it isn’t complete, and isn’t really usable entirely, even with the currently established features (as far I’ve tested, but I need to check with them), the Friends system is nearly complete! [https://github.com/PretendoNetwork/friends](https://github.com/PretendoNetwork/friends) 

As I’m not knowledgeable with this third-party implementation, I can’t say for sure if it’s 1:1 identical in terms of the API (or even, provide an extended API for those features for that matter). So, this is a bit of a vague suggestion to prepare for such thing, as I also don’t know how future-proof the 3DS Friends system will hold up once the official servers are gone for good.

_PS: They also work on Wii U support if you are interested. I’ve rejoined your Support Discord server as well, so you can contact me there too._

### LINKED ISSUE #23: [Request] Add Pretendo support
I want to be able to use this while using Pretendo but it says the bot's friend code is invalid (makes sense because different server), It'd be useful to add support for Pretendo users, plus the fact Nintendo is shutting down the official servers next year would require this addition.

### PR DESCRIPTION
**Pull request by Preloading**
This is a pretty massive PR, that should (mostly) add Pretendo support. This should resolve #10 and it should resolve #23. I am so so so glad that this mostly done.

**Some other modifications:**

- Updated NintendoClients to latest, but SQLAlchemy & Flask-SQLAlchemy are still outdated.
- ~~I probably broke the client, I tried to make it work, but it seemed to inconvenient to set it up, and after someone said it was deprecated, I decided not to test. I believe that it will probably be deprecated, as the web app is much more convenient and easier to use~~ Edit: Decided to test and fixed everything broken with it. Works with nintendo only though.
- Delete friend from the tables if it no longer has any discord accounts linked to it. (only on deletion from consoles page)
- A new page which allows you to choose between creating a pretendo account, or nintendo account

**Migration Steps**

Since this is a pretty massive update, it does require some database changes. I created a handy migration SQL file. **MAKE SURE TO BACKUP THE DATABASE BEFORE RUNNING THE FILE!!!**, you can't easily revert this, so make sure you have a backup in case if something goes wrong. This should prepare the database for pretendo, but to get it to actually run the pretendo backend, you must use your sacrificial 3ds again, and redump the NEX keys, this time in pretendo mode (you can get into pretendo mode by using Nimbus) You will also need to modify your private.py and public.py according to template.private.py & public.py. You will also need to modify whatever you use to start the backend, to start it twice at the same time with the arguments:

`
python backend.py -n nintendo
`
`
python backend.py -n pretendo
`
or
`
python backend.py --network nintendo
`
`
python backend.py --network pretendo
`
whichever is most convenient for you. This should be enough for you to have both Pretendo & Nintendo running! If it's not for some reason, please either comment on this post, or message me on discord (@Preloading)

Oh also, I recommend you do a test of it first, as I do not have the ablility to add activities.write to my thing, so I have no clue whether discord.py works.

**Image Gallery**
Because I'm proud of doing this
image
image
image

~~**VERY BIG NOTE**~~
~~This isn't quite functional yet, due to a presumed bug in Pretendo, where NintendoClients hangs the script. I've gotten it to mostly work, to the point of testing for a short period, but it will hang if left running.~~

Also Jon, in issue #10, it was not as simple as swapping out the url D: (also thank you for the effort you put in pretendo team. Not only did you deal with my stupid questions, and fixing some sorta date time bug which made the code not work, and eventually the bug which prevents this from being finished, thank you for creating and hosting pretendo!)

### CODE REVIEW

**Comment by Preloading** [comment]
Please wait for pretendo to fix the bug that hangs it before merging

**Comment by Preloading** [comment]
Fixes #29 aswell i guess

File: server/backend.py
Comment (spotlightishere) [inline server/backend.py:24]: Could this instead be an [`IntEnum`](https://docs.python.org/3/library/enum.html#enum.IntEnum) or similar? This would allow lines such as the if/else statement to be written as
```python
if network == NetworkIDsToName.nintendo:
     # […]
elif network == NetworkIDsToName.pretendo:
     # […]
```

On a related note, enum members are written in all capitals (e.g. `NINTENDO`).

File: server/discord.py
Comment (spotlightishere) [inline server/discord.py:15]: Can we define this enum in a separate file and import it so that it doesn’t need to be redefined between files?

File: server/server.py
Comment (spotlightishere) [inline server/server.py:57]: Similarly here - many of these helper functions (mapping networks to FCs, reversing enum name to its key definition) could best be in their own file and imported as needed.

**Review by spotlightishere — CHANGES_REQUESTED** [review]
Congratulations on adding support - this was a big undertaking! Looking forward to seeing this merged once the delay is resolved on Pretendo’s side.

Ultimately Delta has the final say here, but I had a few stylistic comments that may help them to review later on.

File: server/sqlite/CREATE.sql
Comment (Preloading) [reply server/sqlite/CREATE.sql:46]: I changed them to be nullable, as in trying to setup 3DS-RPC on my system, and this seemed like the easiest way to fix it failing to create an account. I do agree setting this to a proper value on account creation would be more effective than having it default to null.

File: server/backend.py
Comment (Preloading) [reply server/backend.py:24]: I'll move it to a separate file, and change it to IntEnum (if my code lets me lol), but I can't quite change it to uppercase, as it's used for the database name, and some UI elements, so uppercase it would cause issues. If you look in the code, my casing is all over the place. I normally program in C# and JS, so my casing is coming from there, which isn't best practice for python.

File: server/discord.py
Comment (Preloading) [reply server/discord.py]: Moved it to a new file called networks.py

File: server/backend.py
Comment (Preloading) [reply server/backend.py]: Changed it to a IntEnum, but I don't think I'll change it to uppercase.

File: server/server.py
Comment (Preloading) [reply server/server.py]: Moved them to another file

**Comment by Preloading** [comment]
> Please wait for pretendo to fix the bug that hangs it before merging

It seems to be functioning after adding a horrifying fix, to make NintendoClients send ACK after 3170 years. This isn't permanent, but it should be fine till Pretendo fixes the bug that breaks this the code without this hack. So I'm gonna say, unless I find any other app-breaking issues, it should be fine to merge! (finally!)

**Comment by Preloading** [comment]
never... mind... it's hanging again


nooooooooooooooooo!

**Comment by DanNguyenN** [comment]
Hope this PR getting reviewed. Nintendo network is shut down

**Comment by gllxflr** [comment]
@MCMi460 please merge, most people won't be on Nintendo server because no online games work with it

**Comment by Preloading** [comment]
It looks like the bug that has been breaking this has been fixed, but Pretendo seems to be having some server issues still that affect this a bit, but shouldn't impact usage of this that much

**Comment by DFBR1208** [comment]
Soooo, any idea of when will this be merged?

**Comment by spotlightishere** [comment]
> Soooo, any idea of when will this be merged?

In general, it's not a good idea to ask for an exact date - open source maintainers have their lives to handle, on top of the work of maintaining and developing projects :)

However, within the NSO-RPC Discord server, @MCMi460 stated they would like to begin reviewing this within the coming weeks.

**Comment by DFBR1208** [comment]
> > Soooo, any idea of when will this be merged?
> 
> In general, it's not polite to ask for an exact date - open source maintainers have their lives to handle, on top of the work of maintaining and developing projects :)
> 
> However, within the NSO-RPC Discord server, @MCMi460 stated they would like to begin reviewing this within the coming weeks.

Sorry if it sounded rude, just wanted to know about any update on the project so far, thx for replying!!

**Review by spotlightishere — APPROVED** [review]
All looks good to me!

**Comment by wish13yt** [comment]
One quick question (I am not sure how PRs work really). If this PR got approved, why is it still open?

**Comment by TwistedTempest** [comment]
> One quick question (I am not sure how PRs work really). If this PR got approved, why is it still open?

Once a pull request is tested against the codebase it's supposed to merge with, it can be approved as functional, but still needs to actually be pushed into the codebase by a developer with that level of permissions (which are fairly few and far between in most projects)

**Comment by wish13yt** [comment]
Thanks! I'm not that experienced with how GitHub works.

**Comment by Preloading** [comment]
> One quick question (I am not sure how PRs work really). If this PR got approved, why is it still open?

It is also technically merged (see https://github.com/MCMi460/3DS-RPC/commit/8751376416cdecdc93697052fa157fedec3c6a31), but it was merged pretty poorly, missing some lines. So i really don't know why it's open

**Comment by HotaruBlaze** [comment]
It'll most likely get merged when Pretendo support is working and live.

**Comment by gllxflr** [comment]
@Preloading does that mean it should work then?

**Comment by onyxcode** [comment]
really hyped for this!!

**Review by QuizzaciousMan — APPROVED** [review]
👍

File: server/backend.py
Comment (jonbarrow) [inline server/backend.py:70]: We made a lot of stability changes and bug fixes a while ago, after this PR was made. This hack can likely be removed, and if the issue still persists then you should open an issue about it on our end. This is obviously bad behavior on our end if it persists

Comment (Preloading) [reply server/backend.py:70]: Hi there! A bit ago, I tested, and the program does work without these lines. (yay, thank you!) However, the PR has technically been merged (see https://github.com/MCMi460/3DS-RPC/commit/8751376416cdecdc93697052fa157fedec3c6a31), so I can't modify the PR without causing problems (probably). Why this isn't closed, I do not know. I should probably close it. I'll make sure to try to get the owner of the repo to remove the workaround. Thanks tho!

**Comment by Preloading** [comment]
Squashed and merged in commit https://github.com/MCMi460/3DS-RPC/commit/8751376416cdecdc93697052fa157fedec3c6a31

**Pull request closed without merging**
